### PR TITLE
Adding support for the package attribute on the types element.

### PR DESCRIPTION
**Pull request by ratcashdev**
Implementation of #507 for Java.

The main idea is that we store the package attribute read from the XML to the IR, into token[0]. Then it's up to the generator to make use of that extra piece of information (controllable through `sbe.type.package.override` argument, defaulting to `false` ).

The serialization of the IR is not touched, therefore the package information is not present in the serialized IR.

Some classes in this PR, like the `MultiPackageOutputManager` and `StringWriterOutputManager` actually belong to agrona, but have been implemented here to have a testable code.
Before merging, those files/classes should be moved and substituted with a new release of agrona.

### CODE REVIEW

**Comment by ratcashdev** [comment]
@mjpt777 @tmontgomery Hi, subject to Your time permitting, I'd appreciate a 1st, rough review of this. Thanks.

**Comment by mjpt777** [comment]
Sorry but we have been busy. We need to give priority to our commercial support customers.

**Comment by ratcashdev** [comment]
Sure, take your time. Just wanted to make sure this blipped on your radar.

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/TargetCodeGeneratorLoader.java [hunk 1]
```diff
@@ -52,7 +52,8 @@ public CodeGenerator newInstance(final Ir ir, final String outputDir)
                 "true".equals(System.getProperty(JAVA_GROUP_ORDER_ANNOTATION)),
                 "true".equals(System.getProperty(JAVA_GENERATE_INTERFACES)),
                 "true".equals(System.getProperty(DECODE_UNKNOWN_ENUM_VALUES)),
-                new PackageOutputManager(outputDir, ir.applicableNamespace()));
+                "true".equals(System.getProperty(TYPE_PACKAGE_OVERRIDE)),
```
Comment (mjpt777) [inline sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/TargetCodeGeneratorLoader.java:55]: Should this default to `true`?

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/xml/Type.java [hunk 2]
```diff
@@ -36,7 +38,7 @@
     private int sinceVersion;
 
     /**
-     * Construct a new Type from XML Schema. Called by subclasses to mostly set common fields
```
Comment (mjpt777) [inline sbe-tool/src/main/java/uk/co/real_logic/sbe/xml/Type.java:39]: Accidental delete?

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/java/JavaOutputManager.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import org.agrona.collections.Object2ObjectHashMap;
+import org.agrona.generation.PackageOutputManager;
+import uk.co.real_logic.sbe.generation.MultiPackageOutputManager;
+
+/**
+ * Implementation of {@link MultiPackageOutputManager} for Java.
+ */
+public class JavaOutputManager implements MultiPackageOutputManager
+{
+
```
Comment (mjpt777) [inline sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/java/JavaOutputManager.java:31]: What's with the blank line?

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/java/JavaOutputManager.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    private final String baseDirName;
+    private final PackageOutputManager global;
```
Comment (mjpt777) [inline sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/java/JavaOutputManager.java:33]: Naming should be more explicit.

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/ir/Token.java [hunk 5]
```diff
@@ -105,6 +107,7 @@
     public Token(
         final Signal signal,
         final String name,
+        final String packageName,
```
Comment (mjpt777) [inline sbe-tool/src/main/java/uk/co/real_logic/sbe/ir/Token.java:110]: Should come at end of string args as new.

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/java/JavaOutputManager.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        acting = global;
+    }
+
+    @Override
+    public Writer createOutput(final String name) throws IOException
+    {
+        return new FilterWriter(acting.createOutput(name))
+        {
+            @Override
```
Comment (mjpt777) [inline sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/java/JavaOutputManager.java:71]: We don't apply @Override in code style.

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/TargetCodeGeneratorLoader.java (hunk 1 above)
Comment (ratcashdev) [reply sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/TargetCodeGeneratorLoader.java:55]: unless `TYPE_PACKAGE_OVERRIDE` is explicitly set to `true`, the generator will not respect package overrides, therefore behave as before this change - unless I am missing something here?

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/java/JavaOutputManager.java (hunk 3 above)
Comment (ratcashdev) [reply sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/java/JavaOutputManager.java]: Fixed.

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/xml/Type.java (hunk 2 above)
Comment (ratcashdev) [reply sbe-tool/src/main/java/uk/co/real_logic/sbe/xml/Type.java]: Yes, unfortunately. Reverted.

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/java/JavaOutputManager.java (hunk 4 above)
Comment (ratcashdev) [reply sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/java/JavaOutputManager.java]: Naming's difficult. But tried to do my best.

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/ir/Token.java (hunk 5 above)
Comment (ratcashdev) [reply sbe-tool/src/main/java/uk/co/real_logic/sbe/ir/Token.java]: Done.

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/ir/Token.java [hunk 7]
```diff
@@ -361,6 +375,7 @@ public String toString()
     {
         private Signal signal;
         private String name;
+        private String packageName;
```
[the remark this answers is not in the document]
Comment (ratcashdev) [reply sbe-tool/src/main/java/uk/co/real_logic/sbe/ir/Token.java]: Set the default to null. Also added explanation to the javadoc.

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/java/JavaOutputManager.java (hunk 6 above)
Comment (ratcashdev) [reply sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/java/JavaOutputManager.java]: Right. fixed.

**Comment by ratcashdev** [comment]
Created https://github.com/real-logic/agrona/pull/266 for the required changes in agrona.

**Comment by vyazelenko** [comment]
@ratcashdev Agrona [1.17.0](https://github.com/real-logic/agrona/releases/tag/1.17.0) was released and it contains https://github.com/real-logic/agrona/pull/266.

**Comment by ratcashdev** [comment]
@vyazelenko thx. And I have just realized we i put there a typo: `DynamicPackageOutputManger` instead of `Manager`...

**Comment by ratcashdev** [comment]
updated to latest master.

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/java/JavaOutputManager.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            outputManagerCache.put(packageName, actingPackageOutputManager);
+        }
+    }
+
+    private void resetPackage()
+    {
+        actingPackageOutputManager = basePackageOutputManager;
+    }
+
+    @Override
```
Comment (mjpt777) [inline sbe-tool/src/main/java/uk/co/real_logic/sbe/generation/java/JavaOutputManager.java:65]: @Override not used in code style.

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/xml/XmlSchemaParser.java [hunk 9]
```diff
@@ -326,6 +326,18 @@ public static String getAttributeValue(final Node elementNode, final String attr
...
+    /**
+     * To be used with child elements of {@code <types>} elements. Returns the package attribute as
+     * defined on the parent {@code <types>} element
+     *
+     * @param elementNode the node inside the types element
+     * @return the package name, or null if not defined
+     */
+    public static String getTypesPackageAttribute(final Node elementNode)
+    {
+        return getAttributeValue(elementNode.getParentNode(), "package", null);
```
Comment (mjpt777) [inline sbe-tool/src/main/java/uk/co/real_logic/sbe/xml/XmlSchemaParser.java:338]: This is not likely to be sufficient for nested composites or ref types as the parent will be different.

Comment (ratcashdev) [reply sbe-tool/src/main/java/uk/co/real_logic/sbe/xml/XmlSchemaParser.java:338]: True. I need to write a test case for this. So we speak about two potential cases as below?

```xml
...
<messageSchema package="composite.elements"
               id="3"
               semanticVersion="5.2"
               description="Unit Test"
               byteOrder="littleEndian">
    <types package="specific.package.name">
        <composite name="messageHeader" description="Message identifiers and length of message root">
            <type name="blockLength" primitiveType="uint16"/>
            <type name="templateId" primitiveType="uint16"/>
            <type name="schemaId" primitiveType="uint16"/>
            <type name="version" primitiveType="uint16"/>
        </composite>
      ...
       <!-- case number 1 -->
       <enum name="booleanEnum" encodingType="uint8" semanticType="Boolean">
            <validValue name="F">0</validValue>
            <validValue name="T">1</validValue>
        </enum>
        <composite name="futuresPrice">
            <type name="mantissa" primitiveType="int64" />
            <type name="exponent" primitiveType="int8" />
            <ref name="isSettlement" type="booleanEnum" offset="10" />
        </composite>
 
      <!-- case number 2 -->
      <composite name="outer">
            <enum name="enumOne" description="enum as uint8" encodingType="uint8" semanticType="int">
                <validValue name="Value1">1</validValue>
                <validValue name="Value10">10</validValue>
            </enum>
            <type name="zeroth" primitiveType="uint8"/>
            <set name="setOne" description="set as uint32" encodingType="uint32" semanticType="int">
                <choice name="Bit0">0</choice>
                <choice name="Bit16">16</choice>
                <choice name="Bit26">26</choice>
            </set>
            <composite name="inner">
                <type name="first" primitiveType="int64"/>
                <type name="second" primitiveType="int64"/>
            </composite>
        </composite>
    </types>
...
```

Comment (mjpt777) [reply sbe-tool/src/main/java/uk/co/real_logic/sbe/xml/XmlSchemaParser.java:338]: Yes these cases and consider the likes of: https://github.com/real-logic/simple-binary-encoding/blob/master/sbe-samples/src/main/resources/example-schema.xml#L32-L34

BTW I've pushed a change to better find the ancestor `<types>` element.

**Comment by ratcashdev** [comment]
Ok, will enhance the tests as a separate PR. Thanks for the fix.

**Comment by ratcashdev** [comment]
just realized, the case below also needs to be taken care of (REF pointing to a type in a different package):
```xml
 <types package="specific.package.name.A">
       <enum name="booleanEnum" encodingType="uint8" semanticType="Boolean">
            <validValue name="F">0</validValue>
            <validValue name="T">1</validValue>
        </enum>
</types>

 <types package="specific.package.name.B">
       <composite name="futuresPrice">
            <type name="mantissa" primitiveType="int64" />
            <type name="exponent" primitiveType="int8" />
            <ref name="isSettlement" type="booleanEnum" offset="10" />
        </composite>
</types>
```

**Comment by mjpt777** [comment]
Yes the `<ref>` case needs to be considered.

**Comment by mjpt777** [comment]
When do you think you can address this?

**Comment by ratcashdev** [comment]
Apologies, got sick for the past week, still on meds. Fingers cross - this week.

**Comment by ratcashdev** [comment]
please have a look on https://github.com/real-logic/simple-binary-encoding/pull/915

**Pull request merged**